app/src/main/java/pm/gnosis/heimdall/ui/safe/connect/ConnectExtensionStartActivity.kt

```diff
@@ -0,0 +1,162 @@
+
+    override fun layout() = R.layout.screen_connect_extension_start
+
+    override fun inject(component: ViewComponent) = viewComponent().inject(this)
+
+    override fun screenId() = ScreenId.CONNECT_BROWSER_EXTENSION
+
+    private lateinit var safe: Solidity.Address
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        safe = intent.getStringExtra(EXTRA_SAFE_ADDRESS)?.asEthereumAddress() ?: run {
+            finish()
+            return
+        }
+
+        bottomPanel.disabled = true
+
+        authenticatorLink.apply {
+            setOnClickListener { shareExternalText(BuildConfig.CHROME_EXTENSION_URL, "Browser Extension URL") }
+        }
+
+        viewModel.setup(safe)
+
+        viewModel.observableState.observe(this, Observer {
+
+            onLoading(it.isLoading)
+
+            when(it) {
+
+                is PairingAuthenticatorContract.ViewUpdate.Balance -> {
+
+                    if (it.sufficient) {
+                        bottomPanel.disabled = false
+                        feesError.visibility = View.GONE
+                    } else {
+                        bottomPanel.disabled = true
+                        feesError.text = getString(R.string.insufficient_funds_please_add, it.tokenSymbol)
+                        feesError.visibility = View.VISIBLE
+                    }
+                    progressBar.visibility = View.GONE
+                    balanceBeforeValue.text = it.balanceBefore
+                    feesValue.text = it.fee.underline()
+                    balanceAfterValue.text = it.balanceAfter
+                }
+
+                is PairingAuthenticatorContract.ViewUpdate.Pairing -> {
+                    it.pairingResult?.let {
+                        when (it) {
+                            is PairingAuthenticatorContract.PairingResult.PairingSuccess -> {
+                                toast(R.string.devices_paired_successfuly)
+                                startActivity(
+                                    ReviewTransactionActivity.createIntent(
+                                        context = this,
+                                        safe = intent.getStringExtra(EXTRA_SAFE_ADDRESS).asEthereumAddress()!!,
+                                        txData = TransactionData.ConnectExtension(it.extension)
+                                    )
+                                )
+                            }
+
+                            is PairingAuthenticatorContract.PairingResult.PairingError -> {
+                                snackbar(coordinator, R.string.error_pairing_devices)
+                                Timber.e(it.error)
+                            }
+                        }
+                    }
+                }
+
+            }
+        })
+
+        swipeToRefresh.setOnRefreshListener {
+            viewModel.estimate()
+            swipeToRefresh.isRefreshing = false
```

Comment:
Why is this required?

Reply:
if user decides to refresh and update the gas price estimate

Reply:
ahh you only use pull to refresh to trigger the loading, not to show that it is loading